cmd/scollector/collectors/haproxy_unix.go

```diff
@@ -0,0 +1,518 @@
+package collectors
+
+import (
+	"encoding/csv"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+)
+
+var trackedInstances = []*trackedInstance{}
+
+type trackedInstance struct {
+	Tier     string
+	URL      string
+	Username string
+	Password string
+}
+
+// HAProxy registers an HAProxy collector.
+func HAProxy(user, pwd string) {
+	collectors = append(collectors, &IntervalCollector{
+		F: func() (opentsdb.MultiDataPoint, error) {
+			return cHAProxyCSV(user, pwd)
+		},
+		name: "haproxy",
+	})
+}
+
+// AddHAProxyInstance adds haproxy instances to be tracked.
+func AddHAProxyInstance(conf string) error {
+	sp := strings.SplitN(conf, ":", 2)
+	if len(sp) < 2 {
+		return fmt.Errorf("haproxy_instance requires two fields")
+	}
+	var instance trackedInstance
+	instance.Tier = sp[0]
+	instance.URL = sp[1]
+	trackedInstances = append(trackedInstances, &instance)
+
+	return nil
+}
+
+func cHAProxyCSV(user, pwd string) (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	var err error
+	client := http.Client{}
+
+	for _, instance := range trackedInstances {
+		instance.Username = user
+		instance.Password = pwd
+		if e := instance.fetchHAProxyData(&md, &client); e != nil {
+			err = e
+		}
+	}
+	return md, err
+}
+
+func (instance *trackedInstance) fetchHAProxyData(md *opentsdb.MultiDataPoint, client *http.Client) error {
+	var err error
+	const metric = "haproxy"
+	parse := func(v string) (int64, error) {
+		var i int64
+		if v != "" {
+			i, err = strconv.ParseInt(v, 10, 64)
+			if err != nil {
+				return 0, err
+			}
+			return i, nil
+		}
+		return i, nil
+	}
+
+	req, err := http.NewRequest("GET", instance.URL, nil)
+	if err != nil {
+		return err
+	}
+	req.SetBasicAuth(instance.Username, instance.Password)
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	reader := csv.NewReader(resp.Body)
+	records, err := reader.ReadAll()
+	if err != nil {
+		return err
+	}
+	if len(records) < 2 {
+		return nil
+	}
+	for _, rec := range records[1:] {
+		// There is a trailing comma in haproxy's csv
+		if len(rec) != len(haproxyCSVMeta)+1 {
+			return fmt.Errorf("expected %v lines. got: %v",
+				len(haproxyCSVMeta)+1, len(rec))
+		}
+		hType := haproxyType[rec[32]]
+		pxname := rec[0]
+		svname := rec[1]
+		ts := opentsdb.TagSet{"pxname": pxname, "svname": svname, "tier": instance.Tier}
+		for i, field := range haproxyCSVMeta {
+			m := strings.Join([]string{metric, hType, field.Name}, ".")
+			switch i {
```

Comment:
This might make sense as an if/else chain rather than a switch. Then you can look at more data than just the index for the conditionals and make them make a bit more sense.